src/controllers/restatedeployment/reconcilers/knative.rs

```diff
@@ -0,0 +1,938 @@
+use std::collections::BTreeMap;
+use std::time::Duration;
+
+use kube::api::{Api, Patch, PatchParams};
+use kube::{Resource, ResourceExt};
+use serde_json::json;
+use tracing::*;
+
+use crate::controllers::restatedeployment::controller::{
+    Context, RESTATE_DEPLOYMENT_ID_ANNOTATION,
+};
+use crate::controllers::restatedeployment::reconcilers::replicaset::generate_pod_template_hash;
+use crate::resources::knative::{Configuration, Revision, Route};
+use crate::resources::restatedeployments::{KnativeDeploymentStatus, RestateDeployment};
+use crate::{Error, Result};
+
+const RESTATE_TAG_ANNOTATION: &str = "restate.dev/tag";
+const RESTATE_REGISTERED_AT_ANNOTATION: &str = "restate.dev/registered-at";
+const RESTATE_DEPLOYMENT_ANNOTATION: &str = "restate.dev/deployment";
+
+/// Main Knative reconciliation function
+/// Updates status in-place and returns the next_removal time for requeue logic
+pub async fn reconcile_knative(
+    ctx: &Context,
+    rsd: &RestateDeployment,
+    status: &mut crate::resources::restatedeployments::RestateDeploymentStatus,
+) -> Result<Option<chrono::DateTime<chrono::Utc>>> {
+    let namespace = rsd
```

Comment:
just pass in namespace from the caller, we have logic to get it and default None/"" higher up